### PR TITLE
Dynamic Surrogates for Case Studies 

### PR DESCRIPTION
**Pull request by Xinhe-Chen**
## Addresses issue:
#85 
## Summary/Motivation:
Surrogate automation workflow for different case studies. 
## Changes proposed in this PR:
- Make this new PR to solve some testing issues
- After creating this PR, close PR #98 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md and COPYRIGHT.md file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by Xinhe-Chen** [comment]
Open this new PR to replace PR #98.

**Comment by Xinhe-Chen** [comment]
@lbianchi-lbl Hi Ludovico, I got some errors in the tests. To test my PR, I generated some data in the csv file and put them in the dispatches/tests/test_data folder in this branch. It seems that Github can't locate them correctly. 
If I want to read some data in my tests, what is the correct way that I should do?
image

**Comment by lbianchi-lbl** [comment]
> @lbianchi-lbl Hi Ludovico, I got some errors in the tests. To test my PR, I generated some data in the csv file and put them in the dispatches/tests/test_data folder in this branch. It seems that Github can't locate them correctly. If I want to read some data in my tests, what is the correct way that I should do?

@Xinhe-Chen the files are not being located correctly because their paths are being constructed using `os.getcwd()`, which is fragile. Instead, I'd recommend the following:

- Create a directory `dispatches/workflow/train_market_surrogates/dynamic/tests/data` containing:
  - An empty `__init__.py` file
  - The data files currently stored in `dispatches/tests/test_data`
- This will result in the data files being accessible from the Python package named `dispatches.workflow.train_market_surrogates.dynamic.tests.data`
- Remove the now empty `dispatches/tests/test_data` directory
- In `setup.py`, under the `package_data` keyword, add an entry for the newly created `dispatches.workflow.train_market_surrogates.dynamic.tests.data`
  - This should be a list containing the data files (names only)
  - You should be able to use the existing entries as an example
- Rename `test_Simulation_Data.py` to `test_simulation_data.py`
- Inside the file, rename the fixtures `test_simulation_data` to `sample_simulation_data` and `test_input_data` to `sample_input_data`
  - Why: functions named `test_` have a special meaning for pytest (they are considered test functions). I don't think having fixtures named `test_xyz` would necessarily cause problems, but it's confusing and therefore best avoided
- In the fixtures, use the `importlib.resources.path()` pattern described here: https://github.com/gmlc-dispatches/dispatches/wiki/How-to:-get-the-absolute-path-to-a-data-file#for-data-files-located-in-a-python-package, e.g.:

```py
from pathlib import Path
try:
    # Pyton 3.8+
    from importlib import resources
except ImportError:
    # Python 3.7
    import importlib_resources as resources

import pytest


def _get_data_path(file_name: str, package: str = "dispatches.workflow.train_market_surrogates.dynamic.tests.data") -> Path:
    with resources.path(package, file_name) as p:
        return Path(p)

@pytest.fixture
def sample_simulation_data_path() -> Path:
    return _get_data_path("simdatatest.csv")
```

**Comment by lbianchi-lbl** [comment]
@Xinhe-Chen if you want to speed up the CI runs so that only the tests for this PR run, you can temporarily change the command at https://github.com/gmlc-dispatches/dispatches/blob/fcdc05f68de83b7f9d23964de15a69a4476b2888/.github/workflows/checks.yml#L81 to e.g. `pytest -k train_market_surrogates` so that only the corresponding subset of tests are run.

The current test failures (`TypeError` about `str` being expected) can be resolved by calling `str()` before passing the `pathlib.Path` object where the `str` object is required (e.g. as far as I understand, `TrainNNSurrogates()`).

**Comment by Xinhe-Chen** [comment]
@adowling2 @bknueven This PR is ready for review.

File: dispatches/workflow/train_market_surrogates/dynamic/do_train_surrogates.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. Both files are also available online at the URL:
+# "https://github.com/gmlc-dispatches/dispatches".
+#################################################################################
+
+import os
+from Simulation_Data import SimulationData
+from Train_NN_Surrogates import TrainNNSurrogates
+from Time_Series_Clustering import TimeSeriesClustering
```
Comment (lbianchi-lbl) [inline dispatches/workflow/train_market_surrogates/dynamic/do_train_surrogates.py:18]: These import paths should be absolute (starting with `dispatches.workflow.train_market_surrogates`) or relative (starting with one or more `.`). For more information see e.g. https://realpython.com/absolute-vs-relative-python-imports/

**Review by lbianchi-lbl — CHANGES_REQUESTED** [review]
I've looked at the files and left comments focusing mostly "low-hanging fruits" in terms of software engineering. @Xinhe-Chen Hopefully this doesn't come across as overly critical: this PR is large and likely represents even larger amounts of work from your side, and it's just a matter of keeping iterating on it a few times to make the code easier to deploy. 

A few more general comments:

- As of 9acf0179827348bcf6f300d092a29a6dfd436fb3, there are many `.jpg` files left in this branch. Unless there's a very strong reason to keep them there, they should be removed
- Some Python file names are not conforming to the `snake_case` naming convention (all lowercase letters separated by `_`) . Again, unless there's a strong reason to deviate from this, they should be modified (e.g. `Simulation_Data.py` should be `simulation_data.py`)
- There are several blocks of commented-out code left in the Python files. Pragmatically, this might make sense in some circumstances (e.g. quickly showcasing an alternative when creating a function would be more cumbersome), but we should make sure that any commented code that's not needed is removed

File: dispatches/workflow/train_market_surrogates/dynamic/Train_NN_Surrogates.py [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and license
+# information, respectively. Both files are also available online at the URL:
+# "https://github.com/gmlc-dispatches/dispatches".
+#
+#################################################################################
+import os
+
+__this_file_dir__ = os.getcwd()
+import sys 
+sys.path.append(__this_file_dir__)
```
[the remark this answers is not in the document]
Comment (Xinhe-Chen) [reply dispatches/workflow/train_market_surrogates/dynamic/Train_NN_Surrogates.py:19]: I will check all the imports. For most imports I did copy and paste.

**Comment by Xinhe-Chen** [comment]
Thank you for the suggestions from @lbianchi-lbl . I reviewed again about the codes and I think I have some new things to do. 

- [ ] Optimize the codes that I read or save data. I should use better path tools like pathlib.Path. 
- [ ] Use jupyter notebooks to demonstrate how to train the surrogates instead of using a py script.
- [ ] Add some more tests to the util codes (codes that extract data from original data.)

**Comment by Xinhe-Chen** [comment]
@adowling2 I have generated the min/max/median/max mileage/min mileage profiles. 
I think the improvement is limited. 
FE_dispatch_95_5_mileage_2

For some plots, the max captures the storage usage but 95% quantile does not. 
FE_dispatch_95_5_mileage_14

So I will say we will use the min/max/median as the sub-scenarios will be the best choice.

**Comment by Xinhe-Chen** [comment]
@adowling2 I have generated the min/max/median/max mileage/min mileage profiles. 
I think the improvement is limited. 
FE_dispatch_95_5_mileage_2

For some plots, the max captures the storage usage but 95% quantile does not. 
FE_dispatch_95_5_mileage_14

So I will say we will use the min/max/median as the sub-scenarios will be the best choice.

**Comment by adowling2** [comment]
@Xinhe-Chen In the legend you should list the average capacity factor in ( ) for each line.

Please also see these tips on formatting figures: https://ndcbe.github.io/data-and-computing/notebooks/01/Publication-Quality-Figures.html

**Comment by Xinhe-Chen** [comment]
@adowling2 Thanks for the comments. I have updated the plots. I made the plots larger sizes to show the legends. 

FE_dispatch_95_5_mileage_2

FE_dispatch_95_5_mileage_14

File: setup.py [hunk 3]
```diff
@@ -121,6 +121,9 @@ class SpecialDependencies:
         "graphviz",
         "gridx-prescient>=2.2.1",
         "nrel-pysam>=3.0.1",
+        "tslearn>=0.5.2",
```
Comment (lbianchi-lbl) [inline setup.py:124]: The `dynamic_sweep` data package need to be installed for it to work in CI:

```suggestion
        "dispatches-dynamic-sweep-data @ git+https://github.com/gmlc-dispatches/dynamic-sweep-data@main",
        "tslearn>=0.5.2",
```

**Review by lbianchi-lbl — APPROVED** [review]

**Comment by lbianchi-lbl** [comment]
The only failing check is the test coverage, which is due to the fact that the two relatively large files `Time_Series_Clustering.py` and `Train_NN_Surrogates.py` are only partially covered by tests. I've created #195 to keep track of this.

**Pull request merged**